Don't use cudnnConvolutionBiasActivationForward.

galv opened the pull request:
It supports only the precomputed implicit GEMM implementation of
convolution.

Also fix the bias layout.

galv committed:
Don't use cudnnConvolutionBiasActivationForward.

It supports only the precomputed implicit GEMM implementation of
convolution.

danpovey commented on src/nnet3/convolution-cudnn.cc:
```diff
@@ -232,35 +231,30 @@ void ConvolutionComputation::InitCudnn() {
   // need a dimension of at least 3.
-  int bias_dims[3] = {1, 1, c.num_channels_out};
-  int bias_stride[3] = {c.num_channels_out, c.num_channels_out, 1};
+  int bias_dims[4] = {1, c.num_channels_out, 1, 1};
```
Can you explain - preferably in a comment - why this specific pattern of dims is required? It's mysterious to me.

galv replied:
Done. I can make the bias optional at some point later.

galv committed:
Explain bias dimensions.

galv committed:
Make bias optional.

The ConvolutionComputation class does not know whether or not the bias
is optional at initialization time. It will simply avoid using the bias
if it is a nullptr.

galv commented:
@danpovey I made the bias optional in my latest commit. The ConvolutionComputation object doesn't know whether the bias is optional in my design. It will simply avoid running any bias-related routines if the bias argument points to null. If you want the ConvolutionComputation object to take a boolean specifying whether or not the bias is used, and check that that matches with whether the bias is null or not, feel free to implement it that way.

danpovey commented:
Thanks-- merging.  Try to get in the habit of changing the documentation when you make code changes!  It's easier psychologically to do it right away than later.
